File: set5/2/main.cpp
#include<iostream>
#include<vector>
#include<queue>

using namespace std;

vector<pair<int, int>> G[100001];

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0); 

    int N; cin >> N;
    int A, B, C; cin >> A >> B >> C;
    int M; cin >> M;
    for(int i = 0; i < M; ++i) {
        int u, v, w; cin >> u >> v >> w;
        G[u].emplace_back(v, w);
        G[v].emplace_back(u, w);
    }
    
    priority_queue<pair<int, int>> pq;
    const int INF = 1 << 30;
    vector<int> dist(N + 1, INF);

    for(int x : {A, B, C}) {
        dist[x] = 0;
        pq.emplace(-dist[x], x);
    }
    while(!pq.empty()) {
        auto [d, cur] = pq.top(); pq.pop();
        if(dist[cur] != -d) continue;
        for(auto [nxt, W]: G[cur]) {
            if(dist[nxt] > dist[cur] + W) {
                dist[nxt] = dist[cur] + W;
                pq.emplace(-dist[nxt], nxt);
            }
        }
    }
    
    int mx = 0, ans = 0;
    for(int i = 1; i <= N; ++i) {
        if(mx < dist[i]) {
            mx = dist[i];
            ans = i;
        }
    }
    cout << ans;
    
    return 0;
}

File: set5/2/Main.java
import java.util.*;
import java.lang.*;
import java.io.*;

public class Main {
    final static int INF = 1 << 30;
    static ArrayList<ArrayList<Pair>> G;
    public static void main(String[] args) {
        FastReader rd = new FastReader();

        int N = rd.nextInt();
        int A = rd.nextInt(), B = rd.nextInt(), C = rd.nextInt();
        int M = rd.nextInt();

        G = new ArrayList<>();
        for(int i = 0; i <= N; ++i) G.add(new ArrayList<Pair>());

        for(int i = 0; i < M; ++i) {
            int u = rd.nextInt(), v = rd.nextInt(), w = rd.nextInt();
            G.get(u).add(new Pair(v, w));
            G.get(v).add(new Pair(u, w));
        }

        PriorityQueue<Pair> pq = new PriorityQueue<>();
        int []dist = new int[N + 1];
        for(int i = 1; i <= N; ++i) dist[i] = INF;

        dist[A] = dist[B] = dist[C] = 0;
        pq.add(new Pair(0, A)); pq.add(new Pair(0, B)); pq.add(new Pair(0, C));

        while(!pq.isEmpty()) {
            Pair top = pq.poll();
            int d = top.first;
            int cur = top.second;
            if(dist[cur] != d) continue;
            for(Pair nxt : G.get(cur)) {
                if(dist[nxt.first] > d + nxt.second) {
                    dist[nxt.first] = d + nxt.second;
                    pq.add(new Pair(dist[nxt.first], nxt.first));
                }
            }
        }

        int mx = 0, ans = 0;
        for(int i = 1; i <= N; ++i) {
            if(mx < dist[i]) {
                mx = dist[i];
                ans = i;
            }
        }
        System.out.println(ans);
    }

    static class Pair implements Comparable<Pair> {
        int first, second;
        public Pair() { first = second = 0; }
        public Pair(int a, int b) {
            first = a;
            second = b;
        }
        @Override
        public int compareTo(Pair o) {
            return first - o.first;
        }
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while(st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                }
                catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() { return Integer.parseInt(next()); }
        long nextLong() { return Long.parseLong(next()); }
        double nextDouble() { return Double.parseDouble(next()); }
        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}